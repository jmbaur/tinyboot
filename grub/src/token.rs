use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Ampersand,
    Case,
    CloseBrace,
    CloseBracket,
    Comment(String),
    Do,
    Done,
    Elif,
    Else,
    Esac,
    Fi,
    For,
    Function,
    If,
    Illegal(char),
    In,
    Literal(String),
    Newline,
    OpenBrace,
    Pipe,
    Select,
    Semicolon,
    Then,
    Time,
    Until,
    Value(String),
    While,
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}
