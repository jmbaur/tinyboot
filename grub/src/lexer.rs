use crate::token::Token;
use std::{iter::Peekable, str::Chars};

const LINE_FEED: char = 0xa as char;

pub struct Lexer<'a> {
    current_char: char,
    src: Peekable<Chars<'a>>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            src: input.chars().peekable(),
            current_char: 0 as char,
        }
    }

    fn read_other(&mut self) -> Token {
        let mut other_str = String::from(self.current_char);
        while let Some(&char) = self.src.peek() {
            if char.is_ascii_whitespace() || char == ';' {
                break;
            }

            other_str.push(self.src.next().expect("peek is not None"));
        }

        match other_str.as_str() {
            "case" => Token::Case,
            "do" => Token::Do,
            "done" => Token::Done,
            "elif" => Token::Elif,
            "else" => Token::Else,
            "esac" => Token::Esac,
            "fi" => Token::Fi,
            "for" => Token::For,
            "function" => Token::Function,
            "if" => Token::If,
            "in" => Token::In,
            "select" => Token::Select,
            "then" => Token::Then,
            "time" => Token::Time,
            "until" => Token::Until,
            "while" => Token::While,
            _ => Token::Value(other_str),
        }
    }

    fn read_comment(&mut self) -> Token {
        let mut comment_str = String::new();

        'outer: while let Some(char) = self.src.next() {
            // Determine if this is a multi-line comment.
            'line_feed: {
                if char == LINE_FEED {
                    while let Some(next) = self.src.peek() {
                        // Multi-line comments can have all-whitespace in front of the octothorpes.
                        if next.is_ascii_whitespace() {
                            // Not a multi-line comment since there were two consecutive line feed
                            // characters.
                            if next == &LINE_FEED {
                                break 'outer;
                            }
                            _ = self.src.next().expect("peek is not None");
                            continue;
                        }

                        // Is a multi-line comment, continue building the token value.
                        if next == &'#' {
                            _ = self.src.next().expect("peek is not None");
                            break 'line_feed;
                        }

                        // Not a multi-line comment, finish the comment token.
                        break 'outer;
                    }
                }
            }

            comment_str.push(char);
        }

        Token::Comment(comment_str.trim().to_string())
    }

    fn read_quoted_val(&mut self) -> Token {
        if self.current_char != '"' {
            return Token::Illegal(self.current_char);
        }
        let mut str_val = String::new();
        for char in self.src.by_ref() {
            if char == '"' {
                break;
            }
            str_val.push(char);
        }
        Token::Value(str_val)
    }

    fn read_literal(&mut self) -> Token {
        if self.current_char != '\'' {
            return Token::Illegal(self.current_char);
        }
        let mut str_val = String::new();
        for char in self.src.by_ref() {
            if char == '\'' {
                break;
            }
            str_val.push(char);
        }
        Token::Literal(str_val)
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let Some(char) = self.src.next() else { return None; };
            if char.is_ascii_whitespace() {
                if char == LINE_FEED {
                    return Some(Token::Newline);
                }
                continue;
            }
            self.current_char = char;
            return Some(match self.current_char {
                '"' => self.read_quoted_val(),
                '#' => self.read_comment(),
                '&' => Token::Ampersand,
                ';' => Token::Semicolon,
                '[' => Token::Value(String::from("[")),
                '\'' => self.read_literal(),
                ']' => Token::CloseBracket,
                '{' => Token::OpenBrace,
                '|' => Token::Pipe,
                '}' => Token::CloseBrace,
                _ => self.read_other(),
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tokenize(input: &str) -> Result<Vec<Token>, String> {
        let l = Lexer::new(input);

        let mut tokens = Vec::new();
        for token in l {
            tokens.push(token);
        }

        Ok(tokens)
    }

    #[test]
    fn whitespace() {
        assert_eq!(
            tokenize(
                r#"
                "#
            )
            .unwrap(),
            vec![Token::Newline]
        );
    }

    #[test]
    fn expressions() {
        assert_eq!(
            tokenize("string1 == string2").unwrap(),
            vec![
                Token::Value("string1".to_string()),
                Token::Value("==".to_string()),
                Token::Value("string2".to_string())
            ],
        );
        assert_eq!(
            tokenize("integer1 -gt integer2").unwrap(),
            vec![
                Token::Value("integer1".to_string()),
                Token::Value("-gt".to_string()),
                Token::Value("integer2".to_string()),
            ]
        );
    }

    #[test]
    fn set_command() {
        assert_eq!(
            tokenize("set foo=bar").unwrap(),
            vec![
                Token::Value("set".to_string()),
                Token::Value("foo=bar".to_string()),
            ]
        );
    }

    #[test]
    fn simple_expression() {
        assert_eq!(
            tokenize(
                r#"if [ $default -ne 0 ]; then
                     set default=0
                   fi"#
            )
            .unwrap(),
            vec![
                Token::If,
                Token::Value("[".to_string()),
                Token::Value("$default".to_string()),
                Token::Value("-ne".to_string()),
                Token::Value("0".to_string()),
                Token::CloseBracket,
                Token::Semicolon,
                Token::Then,
                Token::Newline,
                Token::Value("set".to_string()),
                Token::Value("default=0".to_string()),
                Token::Newline,
                Token::Fi,
            ]
        );
    }

    #[test]
    fn menuentry() {
        assert_eq!(
            tokenize("menuentry { linux /path/to/linux; }").unwrap(),
            vec![
                Token::Value("menuentry".to_string()),
                Token::OpenBrace,
                Token::Value("linux".to_string()),
                Token::Value("/path/to/linux".to_string()),
                Token::Semicolon,
                Token::CloseBrace,
            ]
        );
    }

    #[test]
    fn comment() {
        assert_eq!(
            tokenize("foo # bar").unwrap(),
            vec![
                Token::Value("foo".to_string()),
                Token::Comment("bar".to_string()),
            ]
        );

        assert_eq!(
            tokenize(
                r#"# foo
                   # bar

                   # baz"#
            )
            .unwrap(),
            vec![
                Token::Comment("foo\n bar".to_string()),
                Token::Newline,
                Token::Comment("baz".to_string()),
            ]
        );
    }

    #[test]
    fn device_syntax() {
        assert_eq!(
            tokenize("(hd0,1)").unwrap(),
            vec![Token::Value("(hd0,1)".to_string()),]
        );
    }

    #[test]
    fn full_example() {
        let tokens = tokenize(include_str!("../testdata/grub.cfg")).unwrap();
        assert!(!tokens.iter().any(|tok| matches!(tok, Token::Illegal(_))));
    }
}
