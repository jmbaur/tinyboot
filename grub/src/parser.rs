use crate::lexer::Lexer;
use crate::token::Token;
use std::iter::Peekable;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AssignmentStatement {
    pub name: String,
    pub value: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IfStatement {
    pub not: bool,
    pub condition: CommandStatement,
    pub consequence: Vec<Statement>,
    pub elifs: Vec<IfStatement>,
    pub alternative: Vec<Statement>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WhileStatement {
    /// Make the boolean expression evaluate to false in order for the `consequence` to execute.
    pub until: bool,
    pub consequence: Vec<Statement>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum CommandArgument {
    /// A `Value` is one that can have embedded values inside of it, i.e. "${foo}bar" where the value
    /// of `$foo` is embedded inside the string with extra characters. The final value of a value
    /// expression must be obtained by expanding all inner values.
    Value(String),
    /// A `Literal` is one that must be interpreted literally, oppposed to a value expression whose
    /// final value must be expanded. A literal expressions final value is it's initial value.
    Literal(String),
    Block(Vec<Statement>),
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CommandStatement {
    pub command: String,
    pub args: Vec<CommandArgument>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionStatement {
    pub name: String,
    pub body: Vec<Statement>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Statement {
    Assignment(AssignmentStatement),
    Command(CommandStatement),
    Function(FunctionStatement),
    If(IfStatement),
    #[allow(dead_code)]
    While(WhileStatement),
}

#[derive(Debug, PartialEq, Eq)]
pub struct Root {
    pub statements: Vec<Statement>,
}

fn matches_command(s: &str) -> bool {
    matches!(
        s,
        "[" | "acpi"
            | "authenticate"
            | "background_color"
            | "background_image"
            | "badram"
            | "blocklist"
            | "boot"
            | "cat"
            | "chainloader"
            | "clear"
            | "cmosclean"
            | "cmosdump"
            | "cmostest"
            | "cmp"
            | "configfile"
            | "cpuid"
            | "crc"
            | "cryptomount"
            | "cutmem"
            | "date"
            | "devicetree"
            | "distrust"
            | "drivemap"
            | "echo"
            | "eval"
            | "export"
            | "false"
            | "gettext"
            | "gptsync"
            | "halt"
            | "hashsum"
            | "help"
            | "initrd"
            | "initrd16"
            | "insmod"
            | "keystatus"
            | "linux"
            | "linux16"
            | "list_env"
            | "list_trusted"
            | "load_env"
            | "loadfont"
            | "loopback"
            | "ls"
            | "lsfonts"
            | "lsmod"
            | "md5sum"
            | "menuentry"
            | "module"
            | "multiboot"
            | "nativedisk"
            | "normal"
            | "normal_exit"
            | "parttool"
            | "password"
            | "password_pbkdf2"
            | "play"
            | "probe"
            | "rdmsr"
            | "read"
            | "reboot"
            | "regexp"
            | "rmmod"
            | "save_env"
            | "search"
            | "sendkey"
            | "serial"
            | "set"
            | "sha1sum"
            | "sha256sum"
            | "sha512sum"
            | "sleep"
            | "smbios"
            | "source"
            | "submenu"
            | "terminal_input"
            | "terminal_output"
            | "terminfo"
            | "test"
            | "true"
            | "trust"
            | "unset"
            | "verify_detached"
            | "videoinfo"
            | "wrmsr"
            | "xen_hypervisor"
            | "xen_module"
    )
}

pub struct Parser<'a> {
    lexer: Peekable<Lexer<'a>>,
}

impl<'a> Parser<'a> {
    pub fn new(l: Lexer<'a>) -> Self {
        Self {
            lexer: l.peekable(),
        }
    }

    fn must_next_token(&mut self) -> Result<Token, String> {
        self.lexer
            .next()
            .ok_or_else(|| "next token is None".to_string())
    }

    /// Consume newline or semicolon if the peeked token is one of those.
    #[allow(dead_code)]
    fn eat_end(&mut self) -> Result<(), String> {
        if matches!(
            self.lexer.peek(),
            Some(Token::Newline) | Some(Token::Semicolon)
        ) {
            _ = self.must_next_token()?;
        }
        Ok(())
    }

    fn parse_assignment_statement(&mut self, value: String) -> Result<AssignmentStatement, String> {
        let split = value
            .split_once('=')
            .ok_or_else(|| "missing equals in assignment statement".to_string())?;

        let name = split.0.to_string();
        let value = split.1.to_string();
        let value = if value.is_empty() { None } else { Some(value) };

        Ok(AssignmentStatement { name, value })
    }

    fn parse_scope(&mut self) -> Result<Vec<Statement>, String> {
        let mut body = Vec::new();
        loop {
            let next = self.must_next_token()?;
            if next == Token::CloseBrace {
                break;
            }
            if let Some(stmt) = self.parse_statement(next)? {
                body.push(stmt);
            }
        }

        Ok(body)
    }

    fn parse_command_statement(&mut self, command: String) -> Result<CommandStatement, String> {
        let mut args = Vec::new();

        let mut seen_close_bracket = false;
        while let Some(peek_token) = self.lexer.peek() {
            if matches!(peek_token, &Token::Newline | &Token::Semicolon) {
                if command.as_str() == "[" && !seen_close_bracket {
                    return Err(
                        "opening bracket does not have matching closing bracket".to_string()
                    );
                }
                break;
            }

            let token = self.must_next_token()?;
            match token {
                Token::Value(value) => args.push(CommandArgument::Value(value)),
                Token::Literal(literal) => args.push(CommandArgument::Literal(literal)),
                Token::CloseBracket => {
                    if command.as_str() != "[" {
                        return Err(
                            "closing bracket does not have matching opening bracket".to_string()
                        );
                    } else {
                        seen_close_bracket = true;
                        continue;
                    }
                }
                Token::OpenBrace => args.push(CommandArgument::Block(self.parse_scope()?)),
                _ => return Err(format!("invalid syntax: {:?}", token)),
            };
        }

        let command = if command.as_str() == "[" {
            "test".to_string()
        } else {
            command
        };

        Ok(CommandStatement { command, args })
    }

    fn parse_if_statement(&mut self) -> Result<IfStatement, String> {
        self.parse_if_statement_if_or_elif()
    }

    fn parse_if_statement_if_or_elif(&mut self) -> Result<IfStatement, String> {
        let (not, condition) = {
            let next = self.must_next_token()?;
            let not = next == Token::ExclamationPoint;
            let start_condition_token = if not { self.must_next_token()? } else { next };
            let Token::Value(value) = start_condition_token else {
                return Err("if statement missing value token as start of condition".to_string());
            };
            if !matches_command(&value) {
                return Err("if statement condition must be a command statement".to_string());
            }
            (not, self.parse_command_statement(value)?)
        };

        let next = self.must_next_token()?;
        if !matches!(next, Token::Newline | Token::Semicolon) {
            return Err("missing end of condition (newline or semicolon)".to_string());
        }

        let next = self.must_next_token()?;
        if next != Token::Then {
            return Err("missing then in if statement".to_string());
        }

        let mut alternative = Vec::new();
        let mut consequence = Vec::new();
        let mut elifs = Vec::new();

        while let Some(token) = self.lexer.next() {
            match token {
                Token::Elif => elifs.push(self.parse_if_statement_if_or_elif()?),
                Token::Else => alternative = self.parse_if_statement_else()?,
                Token::Fi => break,
                _ => {
                    if let Some(stmt) = self.parse_statement(token)? {
                        consequence.push(stmt);
                    }
                }
            }
        }

        Ok(IfStatement {
            not,
            condition,
            consequence,
            elifs,
            alternative,
        })
    }

    fn parse_if_statement_else(&mut self) -> Result<Vec<Statement>, String> {
        let mut consequence = Vec::new();

        loop {
            let next = self.must_next_token()?;
            match next {
                Token::Fi => break,
                _ => {
                    if let Some(stmt) = self.parse_statement(next)? {
                        consequence.push(stmt);
                    }
                }
            }
        }

        Ok(consequence)
    }

    fn parse_function_statement(&mut self) -> Result<FunctionStatement, String> {
        let Token::Value(name) = self.must_next_token()? else {
            return Err("function does not have a name".to_string());
        };
        if self.must_next_token()? != Token::OpenBrace {
            return Err("function does not have opening brace".to_string());
        }
        let body = self.parse_scope()?;
        Ok(FunctionStatement { name, body })
    }

    fn parse_statement(&mut self, start_token: Token) -> Result<Option<Statement>, String> {
        Ok(match start_token {
            Token::Newline | Token::Semicolon | Token::Comment(_) => None,
            Token::If => Some(Statement::If(self.parse_if_statement()?)),
            Token::While | Token::Until => todo!(),
            Token::Function => Some(Statement::Function(self.parse_function_statement()?)),
            Token::Value(value) => {
                if matches_command(&value) {
                    Some(Statement::Command(self.parse_command_statement(value)?))
                } else {
                    Some(Statement::Assignment(
                        self.parse_assignment_statement(value)?,
                    ))
                }
            }
            _ => return Err(format!("invalid syntax: {:?}", start_token)),
        })
    }

    pub fn parse(&mut self) -> Result<Root, String> {
        let mut root = Root { statements: vec![] };

        while let Some(token) = self.lexer.next() {
            if let Some(stmt) = self.parse_statement(token)? {
                root.statements.push(stmt);
            }
        }

        Ok(root)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn assert_assignment_statement(
        stmt: &Statement,
        name: impl Into<String>,
        value: Option<impl Into<String>>,
    ) {
        let Statement::Assignment(assignment) = stmt else {
                 panic!("not an assignment statement");
        };
        assert_eq!(assignment.name, name.into());
        assert_eq!(assignment.value, value.map(|s| s.into()));
    }

    fn assert_command_statement(
        stmt: &Statement,
        command: impl Into<String>,
        args: Vec<CommandArgument>,
    ) {
        let Statement::Command(stmt) = stmt else {
            panic!("not a command expression statement");
        };
        assert_eq!(stmt.command, command.into());
        assert_eq!(stmt.args, args);
    }

    fn assert_if_statement(
        stmt: &Statement,
        not: bool,
        condition: CommandStatement,
        consequence: Vec<Statement>,
        elifs: Vec<IfStatement>,
        alternative: Vec<Statement>,
    ) {
        let Statement::If(if_stmt) = stmt else {
            panic!("not an if statement");
        };
        assert_eq!(if_stmt.not, not);
        assert_eq!(if_stmt.condition, condition);
        assert_eq!(if_stmt.consequence, consequence);
        assert_eq!(if_stmt.elifs, elifs);
        assert_eq!(if_stmt.alternative, alternative);
    }

    fn assert_function_statement(stmt: &Statement, name: impl Into<String>, body: Vec<Statement>) {
        let Statement::Function(function) = stmt else {
            panic!("not a function statement");
        };
        assert_eq!(function.name, name.into());
        assert_eq!(function.body, body);
    }

    #[test]
    fn test_assignment_statement() {
        let l = Lexer::new(
            r#"foo=bar
               bar="#,
        );
        let mut p = Parser::new(l);
        let root = p.parse().unwrap();
        assert!(root.statements.len() == 2);
        assert_assignment_statement(&root.statements[0], "foo", Some("bar"));
        assert_assignment_statement(&root.statements[1], "bar", None::<String>);
    }

    #[test]
    fn test_command_statement() {
        let mut p = Parser::new(Lexer::new(r#"[ "${grub_platform}" = "efi" ]"#));
        let root = p.parse().unwrap();
        assert!(root.statements.len() == 1);
        assert_command_statement(
            &root.statements[0],
            "test",
            vec![
                CommandArgument::Value("${grub_platform}".to_string()),
                CommandArgument::Value("=".to_string()),
                CommandArgument::Value("efi".to_string()),
            ],
        );
    }

    #[test]
    fn test_multiple_command_statements() {
        let mut p = Parser::new(Lexer::new("load_env; insmod foo 'bar'"));
        let root = p.parse().unwrap();
        assert!(root.statements.len() == 2);
        assert_command_statement(&root.statements[0], "load_env", vec![]);
        assert_command_statement(
            &root.statements[1],
            "insmod",
            vec![
                CommandArgument::Value("foo".to_string()),
                CommandArgument::Literal("bar".to_string()),
            ],
        );
    }

    #[test]
    fn test_full_if_statement() {
        let mut p = Parser::new(Lexer::new(
            r#"if [ "foo" ]; then; elif test "bar"; then; else; fi"#,
        ));
        let root = p.parse().unwrap();
        assert!(root.statements.len() == 1);
        assert_if_statement(
            &root.statements[0],
            false,
            CommandStatement {
                command: "test".to_string(),
                args: vec![CommandArgument::Value("foo".to_string())],
            },
            vec![],
            vec![IfStatement {
                not: false,
                condition: CommandStatement {
                    command: "test".to_string(),
                    args: vec![CommandArgument::Value("bar".to_string())],
                },
                consequence: vec![],
                alternative: vec![],
                elifs: vec![],
            }],
            vec![],
        );
    }

    #[test]
    fn test_function() {
        // TODO(jared): implement function calls
        let src = r#"
            function foobar { load_env; }
            # foobar "foo" "bar"
        "#;
        let mut p = Parser::new(Lexer::new(src));
        let root = p.parse().unwrap();
        assert!(root.statements.len() == 1);
        assert_function_statement(
            &root.statements[0],
            "foobar",
            vec![Statement::Command(CommandStatement {
                command: "load_env".to_string(),
                args: vec![],
            })],
        );
    }

    #[test]
    fn test_full_example() {
        let mut p = Parser::new(Lexer::new(include_str!("../testdata/grub.cfg")));
        p.parse().expect("no parsing errors");
    }
}
