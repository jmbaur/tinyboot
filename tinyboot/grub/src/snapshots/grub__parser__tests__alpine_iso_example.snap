---
source: grub/src/parser.rs
expression: ast
---
Root {
    statements: [
        Command(
            CommandStatement {
                command: "set",
                args: [
                    Value(
                        "timeout=1",
                    ),
                ],
            },
        ),
        Command(
            CommandStatement {
                command: "menuentry",
                args: [
                    Value(
                        "Linux lts",
                    ),
                    Block(
                        [
                            Command(
                                CommandStatement {
                                    command: "linux",
                                    args: [
                                        Value(
                                            "/boot/vmlinuz-lts",
                                        ),
                                        Value(
                                            "modules=loop,squashfs,sd-mod,usb-storage",
                                        ),
                                        Value(
                                            "quiet",
                                        ),
                                    ],
                                },
                            ),
                            Command(
                                CommandStatement {
                                    command: "initrd",
                                    args: [
                                        Value(
                                            "/boot/initramfs-lts",
                                        ),
                                    ],
                                },
                            ),
                        ],
                    ),
                ],
            },
        ),
    ],
}
